api/storage/deployment_storage.go

```diff
@@ -83,3 +89,84 @@ func (d *deploymentStorage) GetFirstSuccessModelVersionPerModel() (map[models.ID
 func (d *deploymentStorage) Delete(modelID models.ID, versionID models.ID) error {
 	return d.db.Where("version_id = ? AND version_model_id = ?", versionID, modelID).Delete(models.Deployment{}).Error
 }
+
+// OnDeploymentSuccess updates the new deployment status to successful on DB and update all previous deployment status for that version endpoint to terminated.
+func (d *deploymentStorage) OnDeploymentSuccess(newDeployment *models.Deployment) error {
+	if newDeployment.ID == 0 {
+		return fmt.Errorf("newDeployment.ID must not be empty")
+	}
+
+	if newDeployment.Status != models.EndpointRunning && newDeployment.Status != models.EndpointServing {
+		return fmt.Errorf("newDeployment.Status must be running or serving")
+	}
+
+	tx := d.db.Begin()
+	if tx.Error != nil {
+		return fmt.Errorf("failed to begin transaction: %s", tx.Error)
+	}
+
+	var err error
+	defer func() {
+		if err != nil {
+			tx.Rollback()
+		} else {
+			tx.Commit()
+		}
+	}()
+
+	var deployments []*models.Deployment
+	err = tx.Where("version_model_id = ? AND version_id = ? AND version_endpoint_id = ?",
+		newDeployment.VersionModelID, newDeployment.VersionID, newDeployment.VersionEndpointID).Find(&deployments).Error
```

Comment:
We should only update previous "running" / "serving" endpoints to "terminated", right?

Can we include the filter in this query? If not, potentially, this can also be expensive as it will end up returning all previous deployment attempts. 

Once we do the ad-hoc DB migration, at any given time, there should only be a single deployment record in the DB for a given endpoint that is in the running/serving state. All others will be failed/terminated and can be ignored.

Reply:
We were thinking the same thing. I realized this after doing some testing and am surprised that you were suggesting the same improvement. Thanks for the suggestion.